File: Leetcode 460-LFU cache/lfu_cache.py
from collections import defaultdict
from email import header
class LFUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.size = 0
        self.cache = {}
        self.frequencyMap = defaultdict(LinkedList)
        self.minFreq = 0
        

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self.cache:
            return -1
        
        tempNode = self.cache[key]
        self.frequencyMap[tempNode.freq].removeNode(tempNode)
        if self.frequencyMap[tempNode.freq].head is None:
            del self.frequencyMap[tempNode.freq]
            if self.minFreq == tempNode.freq:
                self.minFreq += 1
        
        self.cache[key].freq += 1
        self.frequencyMap[self.cache[key].freq].insertAtTail(self.cache[key])
        return self.cache[key].value
   
            
    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        
        if self.capacity == 0:
            return
        
        if key in self.cache:
            self.get(key)
            self.cache[key].value = value
            return
        
        if self.size == self.capacity:
            del self.cache[self.frequencyMap[self.minFreq].head.key]
            self.frequencyMap[self.minFreq].removeHead()
            if self.frequencyMap[self.minFreq].head is None:
                del self.frequencyMap[self.minFreq]
            self.size -= 1
        
        self.minFreq = 1
        self.cache[key] = LinkedListNode(key, value, self.minFreq)
        self.frequencyMap[self.cache[key].freq].insertAtTail(self.cache[key])
        self.size += 1
    
    def printFrequencyMap(self):
        print("frequncyMap:")
        for freq, linkedList in self.frequencyMap.items():
            print("frequency", freq)
            print("linked list", printLinkedList(linkedList))

    

def printLinkedList(linkedList):
    current = linkedList.head
    nodes = []
    while current is not None:
        nodes.append(str(current.value))
        current = current.next
    return " -> ".join(nodes)
    
    
class LinkedListNode:
    def __init__(self, key, value, freq):
        self.key = key
        self.value = value
        self.freq = freq
        self.next = None
        self.prev = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
    
    def insertAtTail(self, node):
        node.prev, node.next = None, None
        if self.tail is None:
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            node.prev = self.tail
            self.tail = node
    
    def removeNode(self, node):
        if node is None:
            return
        
        if node.prev is not None:
            node.prev.next = node.next
        if node.next is not None:
            node.next.prev = node.prev
        if node == self.head:
            self.head = self.head.next
        if node == self.tail:
            self.tail = self.tail.prev
    
    def removeHead(self):
        self.removeNode(self.head)
    
if __name__ == "__main__":
    lfuCache = LFUCache(2)
    lfuCache.put(1, 1)
    print("put(1, 1)")
    print("cache", lfuCache.cache)
    lfuCache.printFrequencyMap()

    lfuCache.put(2, 2)
    print("\nput(2, 2)")
    print("cache", lfuCache.cache)
    lfuCache.printFrequencyMap()

    print("\nget(1)")
    print("value", lfuCache.get(1))
    print("cache", lfuCache.cache)
    lfuCache.printFrequencyMap()

    lfuCache.put(3, 3)
    print("\nput(3, 3)")
    print("cache", lfuCache.cache)
    lfuCache.printFrequencyMap()

    print("\nget(2)")
    print("value", lfuCache.get(2))
    print("cache", lfuCache.cache)
    lfuCache.printFrequencyMap()

    print("\nget(3)")
    print("value", lfuCache.get(3))
    print("cache", lfuCache.cache)
    lfuCache.printFrequencyMap()

    lfuCache.put(4, 4)
    print("\nput(4, 4)")
    print("cache", lfuCache.cache)
    lfuCache.printFrequencyMap()

    print("\nget(1)")
    print("value", lfuCache.get(1))
    print("cache", lfuCache.cache)
    lfuCache.printFrequencyMap()

    print("\nget(3)")
    print("value", lfuCache.get(3))
    print("cache", lfuCache.cache)
    lfuCache.printFrequencyMap()

    print("\nget(4)")
    print("value", lfuCache.get(4))
    print("cache", lfuCache.cache)
    lfuCache.printFrequencyMap()



"""
output: 
put(1, 1)
('cache', {1: <__main__.LinkedListNode instance at 0x10e7237e8>})
frequncyMap:
('frequency', 1)
('linked list', '1')

put(2, 2)
('cache', {1: <__main__.LinkedListNode instance at 0x10e7237e8>, 2: <__main__.LinkedListNode instance at 0x10e723830>})
frequncyMap:
('frequency', 1)
('linked list', '1 -> 2')

get(1)
('value', 1)
('cache', {1: <__main__.LinkedListNode instance at 0x10e7237e8>, 2: <__main__.LinkedListNode instance at 0x10e723830>})
frequncyMap:
('frequency', 1)
('linked list', '2')
('frequency', 2)
('linked list', '1')

put(3, 3)
('cache', {1: <__main__.LinkedListNode instance at 0x10e7237e8>, 3: <__main__.LinkedListNode instance at 0x10e723758>})
frequncyMap:
('frequency', 1)
('linked list', '3')
('frequency', 2)
('linked list', '1')

get(2)
('value', -1)
('cache', {1: <__main__.LinkedListNode instance at 0x10e7237e8>, 3: <__main__.LinkedListNode instance at 0x10e723758>})
frequncyMap:
('frequency', 1)
('linked list', '3')
('frequency', 2)
('linked list', '1')

get(3)
('value', 3)
('cache', {1: <__main__.LinkedListNode instance at 0x10e7237e8>, 3: <__main__.LinkedListNode instance at 0x10e723758>})
frequncyMap:
('frequency', 2)
('linked list', '1 -> 3')

put(4, 4)
('cache', {3: <__main__.LinkedListNode instance at 0x10e723758>, 4: <__main__.LinkedListNode instance at 0x10e7237e8>})
frequncyMap:
('frequency', 1)
('linked list', '4')
('frequency', 2)
('linked list', '3')

get(1)
('value', -1)
('cache', {3: <__main__.LinkedListNode instance at 0x10e723758>, 4: <__main__.LinkedListNode instance at 0x10e7237e8>})
frequncyMap:
('frequency', 1)
('linked list', '4')
('frequency', 2)
('linked list', '3')

get(3)
('value', 3)
('cache', {3: <__main__.LinkedListNode instance at 0x10e723758>, 4: <__main__.LinkedListNode instance at 0x10e7237e8>})
frequncyMap:
('frequency', 1)
('linked list', '4')
('frequency', 3)
('linked list', '3')

get(4)
('value', 4)
('cache', {3: <__main__.LinkedListNode instance at 0x10e723758>, 4: <__main__.LinkedListNode instance at 0x10e7237e8>})
frequncyMap:
('frequency', 2)
('linked list', '4')
('frequency', 3)
('linked list', '3')
"""

File: Leetcode 460-LFU cache/lfu_cache.go
package main

import (
	"fmt"
	"strings"
)

type LFUCache struct {
	capacity     int
	size         int
	minFreq      int
	cache        map[int]*LinkedListNode
	frequencyMap map[int]*LinkedList
}

func Constructor(capacity int) LFUCache {
	cache := make(map[int]*LinkedListNode)
	frequencyMap := make(map[int]*LinkedList)
	return LFUCache{
		capacity:     capacity,
		size:         0,
		minFreq:      0,
		cache:        cache,
		frequencyMap: frequencyMap,
	}
}

func (this *LFUCache) Get(key int) int {
	if _, found := this.cache[key]; !found {
		return -1
	}

	tempNode := this.cache[key]
	this.frequencyMap[tempNode.freq].removeNode(tempNode)
	if this.frequencyMap[tempNode.freq].head == nil {
		delete(this.frequencyMap, tempNode.freq)
		if this.minFreq == tempNode.freq {
			this.minFreq += 1
		}
	}

	this.cache[key].freq += 1
	this.initLinkedList(this.cache[key].freq)
	this.frequencyMap[this.cache[key].freq].insertAtTail(this.cache[key])
	return this.cache[key].value
}

func (this *LFUCache) Put(key int, value int) {
	if this.capacity == 0 {
		return
	}

	if _, found := this.cache[key]; found {
		this.Get(key)
		this.cache[key].value = value
		return
	}

	if this.size == this.capacity {
		delete(this.cache, this.frequencyMap[this.minFreq].head.key)
		this.frequencyMap[this.minFreq].removeHead()
		if this.frequencyMap[this.minFreq].head == nil {
			delete(this.frequencyMap, this.minFreq)
		}
		this.size -= 1
	}

	this.minFreq = 1
	this.cache[key] = &LinkedListNode{
		key:   key,
		value: value,
		freq:  this.minFreq,
	}

	this.initLinkedList(this.cache[key].freq)
	this.frequencyMap[this.cache[key].freq].insertAtTail(this.cache[key])
	this.size += 1
}

func (this *LFUCache) initLinkedList(freq int) {
	if _, found := this.frequencyMap[freq]; !found {
		this.frequencyMap[freq] = &LinkedList{}
	}
}

func (this *LFUCache) printFrequencyMap() {
	fmt.Println("frequencyMap: ")
	for freq, linkedList := range this.frequencyMap {
		fmt.Println("frequency: ", freq)
		fmt.Println("linked list", printLinkedList(linkedList))
	}
}

func printLinkedList(linkedList *LinkedList) string {
	current := linkedList.head
	nodes := []string{}
	for current != nil {
		nodes = append(nodes, fmt.Sprintf("%d", current.value))
		current = current.next
	}
	return strings.Join(nodes, " -> ")
}

type LinkedListNode struct {
	key   int
	value int
	freq  int
	next  *LinkedListNode
	prev  *LinkedListNode
}

type LinkedList struct {
	head *LinkedListNode
	tail *LinkedListNode
}

func (ll *LinkedList) insertAtTail(node *LinkedListNode) {
	node.prev, node.next = nil, nil
	if ll.tail == nil {
		ll.head = node
		ll.tail = node
	} else {
		ll.tail.next = node
		node.prev = ll.tail
		ll.tail = node
	}
}

func (ll *LinkedList) removeNode(node *LinkedListNode) {
	if node == nil {
		return
	}
	if node.prev != nil {
		node.prev.next = node.next
	}
	if node.next != nil {
		node.next.prev = node.prev
	}
	if node == ll.head {
		ll.head = ll.head.next
	}
	if node == ll.tail {
		ll.tail = ll.tail.prev
	}
}

func (ll *LinkedList) removeHead() {
	ll.removeNode(ll.head)
}

func main() {
	lfuCache := Constructor(2)
	lfuCache.Put(1, 1)
	fmt.Println("Put(1, 1)")
	fmt.Println("cache: ", lfuCache.cache)
	lfuCache.printFrequencyMap()

	lfuCache.Put(2, 2)
	fmt.Println("\nPut(2, 2)")
	fmt.Println("cache: ", lfuCache.cache)
	lfuCache.printFrequencyMap()

	fmt.Println("\nGet(1)")
	fmt.Println("value", lfuCache.Get(1))
	fmt.Println("cache: ", lfuCache.cache)
	lfuCache.printFrequencyMap()

	lfuCache.Put(3, 3)
	fmt.Println("\nPut(3, 3)")
	fmt.Println("cache: ", lfuCache.cache)
	lfuCache.printFrequencyMap()

	fmt.Println("\nGet(2)")
	fmt.Println("value", lfuCache.Get(2))
	fmt.Println("cache: ", lfuCache.cache)
	lfuCache.printFrequencyMap()

	fmt.Println("\nGet(3)")
	fmt.Println("value", lfuCache.Get(3))
	fmt.Println("cache: ", lfuCache.cache)
	lfuCache.printFrequencyMap()

	lfuCache.Put(4, 4)
	fmt.Println("\nPut(4, 4)")
	fmt.Println("cache: ", lfuCache.cache)
	lfuCache.printFrequencyMap()

	fmt.Println("\nGet(1)")
	fmt.Println("value", lfuCache.Get(1))
	fmt.Println("cache: ", lfuCache.cache)
	lfuCache.printFrequencyMap()

	fmt.Println("\nGet(3)")
	fmt.Println("value", lfuCache.Get(3))
	fmt.Println("cache: ", lfuCache.cache)
	lfuCache.printFrequencyMap()

	fmt.Println("\nGet(4)")
	fmt.Println("value", lfuCache.Get(4))
	fmt.Println("cache: ", lfuCache.cache)
	lfuCache.printFrequencyMap()
}

/* output:
Put(1, 1)
cache:  map[1:0xc0000701e0]
frequencyMap:
frequency:  1
linked list 1

Put(2, 2)
cache:  map[1:0xc0000701e0 2:0xc000070240]
frequencyMap:
frequency:  1
linked list 1 -> 2

Get(1)
value 1
cache:  map[1:0xc0000701e0 2:0xc000070240]
frequencyMap:
frequency:  1
linked list 2
frequency:  2
linked list 1

Put(3, 3)
cache:  map[1:0xc0000701e0 3:0xc000070390]
frequencyMap:
frequency:  1
linked list 3
frequency:  2
linked list 1

Get(2)
value -1
cache:  map[1:0xc0000701e0 3:0xc000070390]
frequencyMap:
frequency:  2
linked list 1
frequency:  1
linked list 3

Get(3)
value 3
cache:  map[1:0xc0000701e0 3:0xc000070390]
frequencyMap:
frequency:  2
linked list 1 -> 3

Put(4, 4)
cache:  map[3:0xc000070390 4:0xc000070570]
frequencyMap:
frequency:  1
linked list 4
frequency:  2
linked list 3

Get(1)
value -1
cache:  map[3:0xc000070390 4:0xc000070570]
frequencyMap:
frequency:  1
linked list 4
frequency:  2
linked list 3

Get(3)
value 3
cache:  map[3:0xc000070390 4:0xc000070570]
frequencyMap:
frequency:  1
linked list 4
frequency:  3
linked list 3

Get(4)
value 4
cache:  map[3:0xc000070390 4:0xc000070570]
frequencyMap:
frequency:  2
linked list 4
frequency:  3
linked list 3
*/
